EIPS/eip-draft-whisper.md

```diff
@@ -0,0 +1,179 @@
+|       | Messages                   |     1     |
+|       | PoW Requirement            |     2     |
+|       | Bloom Filter               |     3     |
+|-------|----------------------------|-----------|
+
+
+The following message codes are optional, but they are reserved for specific purpose.
+
+| EIP   | Name                       | Int Value |
+|-------|----------------------------|-----------|
+|       | P2P Request                |    126    |
+|       | P2P Message                |    127    |
+|-------|----------------------------|-----------|
+
+### Packet Format and Usage
+
+**Status** [`0`, `whisper_protocol_version`] 
+
+This packet contains two objects: integer (0x00) followed by integer (Whisper protocol version).
+
+Informs a peer of the Whisper version. This message should be send after the initial handshake and prior to any other messages.
+
+**Messages** [`1`, `whisper_envelopes`]
+
+This packet contains two objects: integer (0x01) followed by a list of Whisper Envelopes.
+
+This packet is used for sending the standard Whisper envelopes.
+
+**PoW Requirement** [`2`, `PoW`]
+
+This packet contains two objects: integer (0x02) followed by a single floating point value of PoW.
+
+This packet is used by Whisper nodes for dynamic adjustment of their individual PoW requirements. Receipient of this message should no longer deliver the sender messages with PoW lower than specified in this message.
+
+PoW is defined as average number of iterations, required to find the current BestBit (the number of leading zero bits in the hash), divided by message size and TTL:
+
+	PoW = (2**BestBit) / (size * TTL)
+
+PoW calculation:
+
+	fn short_rlp(envelope) = rlp of envelope, excluding env_nonce field.
+	fn pow_hash(envelope, env_nonce) = sha3(short_rlp(envelope) ++ env_nonce)
+	fn pow(pow_hash, size, ttl) = 2**leading_zeros(pow_hash) / (size * ttl)
+
+where size is the size of the full RLP-encoded envelope.
+
+**Bloom Filter** [`3`, `bytes`]
+
+This packet contains two objects: integer (0x03) followed by a byte array of arbitrary size.
+
+This packet is used by Whisper nodes for sharing their interest in messages with specific topics.
+
+The Bloom filter is used to identify a number of topics to a peer without compromising (too much) privacy over precisely what topics are of interest. Precise control over the information content (and thus efficiency of the filter) may be maintained through the addition of bits.
+
+Blooms are formed by the bitwise OR operation on a number of bloomed topics. The bloom function takes the topic and projects them onto a 512-bit slice. At most, three bits are marked for each bloomed topic.
+
+The projection function is defined as a mapping from a a 4-byte slice S to a 512-bit slice D; for ease of explanation, S will dereference to bytes, whereas D will dereference to bits.
+
+	LET D[*] = 0
+	FOREACH i IN { 0, 1, 2 } DO
+	LET n = S[i]
+	IF S[3] & (2 ** i) THEN n += 512
```

Comment:
wouldn't it be `n += 256`, since it's meant to be treated as the 9th bit? adding 512 would overflow, but also be congruent mod 512 in the first place.